src/main/java/backend/resource/TurboLabel.java

```diff
@@ -4,153 +4,185 @@
-            return Optional.empty();
-        }
+    /**
+     * Matches label name that contains groupQuery and the matching is
+     * case-insensitive
+     * @param repoLabels
+     * @param groupQuery
+     * @return the list of label names that belongs to a group and matches a query
+     */
+    public static List<TurboLabel> filterByGroupQuery(List<TurboLabel> repoLabels, String groupQuery) {
+        if (groupQuery.isEmpty()) return repoLabels;
+
+        return repoLabels
+                .stream()
+                .filter(label -> {
+                    if (label.isInGroup()) {
+                        return Utility.containsIgnoreCase(label.getGroupName(), groupQuery);
+                    }
+                    return false;
+                })
+                .collect(Collectors.toList());
     }
 
-    public String getName() {
-        if (getDelimiter(actualName).isPresent()) {
-            String delimiter = getDelimiter(actualName).get();
-            // Escaping due to constants not being valid regexes
-            String[] segments = actualName.split("\\" + delimiter);
-            assert segments.length >= 1;
-            if (segments.length == 1) {
-                if (actualName.endsWith(delimiter)) {
-                    // group.
-                    return "";
-                } else {
-                    // .name
-                    return segments[0];
-                }
-            } else {
-                // group.name
-                assert segments.length == 2;
-                return segments[1];
-            }
-        } else {
-            // name
-            return actualName;
-        }
+    /**
+     * @param repoLabels
+     * @param keyword
+     * @return the list of labels that matches the keyword i.e. the label's group 
+     * contains keyword's group and label's name contains keyword's name
+     */
+    public static List<TurboLabel> getMatchedLabels(List<TurboLabel> repoLabels, String keyword) {
+        String[] extractedNames = splitKeyword(keyword);
+
+        List<TurboLabel> newMatchedLabels = new ArrayList<>();
+        newMatchedLabels.addAll(repoLabels);
+        newMatchedLabels = filterByNameQuery(newMatchedLabels, extractedNames[2]);
+        newMatchedLabels = filterByGroupQuery(newMatchedLabels, extractedNames[0]);
+        return newMatchedLabels;
+    }
+
+    /**
+     * A label is matching if:
+     * the label's group contains keyword's group and label's name contains keyword's name
+     * @param repoLabels
+     * @param keyword
+     * @return true if there is there is at least 1 matching label for the keyword
+     */
+    public static final boolean hasMatchedLabel(List<TurboLabel> repoLabels, String keyword) {
+        return !getMatchedLabels(repoLabels, keyword).isEmpty();
+    }
+
+    /**
+     * @return group name if the label belongs to a group. Empty string if otherwise
+     */
+    public String getGroupName() {
+        return groupName;
+    }
```

Comment:
What was the reason for not returning a `Optional<String>`? Perhaps we can document it?